tools/AutomationTestFramework/Management/ModuleUploader.ps1

```diff
@@ -0,0 +1,98 @@
+﻿# ----------------------------------------------------------------------------------
+    # Copyright Microsoft Corporation
+    # Licensed under the Apache License, Version 2.0 (the "License");
+    # you may not use this file except in compliance with the License.
+    # You may obtain a copy of the License at
+    # http://www.apache.org/licenses/LICENSE-2.0
+    # Unless required by applicable law or agreed to in writing, software
+    # distributed under the License is distributed on an "AS IS" BASIS,
+    # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    # See the License for the specific language governing permissions and
+    # limitations under the License.
+# ----------------------------------------------------------------------------------
+
+function Upload-ModuleToStorage (
+    [hashtable] $storage,
```

Comment:
just curious, why a hashtable instead of a concrete type for storage account (or an IStorageContext)?

Reply:
There are variables in `RunAutomationTests.ps1` like this that are defined as hashtables of mainly creds for these things. So, there is one for storage, one for automation, and another for runbook template related stuff. It is simply so that I don't have to make the functions take 10+ parameters. Plus, I make no classes in any of this scripting code (as I don't really understand/trust PS classes). So, they are just dynamic DTOs so we have the option to split out/parameterize the automation processes later if we need to.

Reply:
Generally, it is preferable to use concrete types rather than hashtables, unless the data is truly user-defined, as this makes it easier on the user to determine what properties they must provide.  For now, hashtables should be OK

Reply:
From what I've seen, internal scripts use a lot of hashtables since it is a simple `@{}` syntax. All functions taking hashtables as input are internal to this process. Obviously, if these were made into individual cmdlets, hashtable isn't a good mechanism for input. Basically, anything in the Management folder is not supposed to be used externally, and only supposed to be used by RunAutomationTests.ps1.